aten/src/ATen/cpu/vec/vec512/vec512_bfloat16.h

```diff
@@ -0,0 +1,848 @@
+  return _mm512_permutexvar_epi64(idx, t_lo);
+}
+
+static inline __m512i merge_compare_result(const __m512& a, const __m512& b) {
+  __m512i lo = _mm512_castps_si512(a);
+  __m512i hi = _mm512_castps_si512(b);
+  lo = _mm512_srli_epi32(lo, 16);
+  hi = _mm512_srli_epi32(hi, 16);
+  auto out = _mm512_packus_epi32(lo, hi);
+  __m512i idx = _mm512_set_epi64(7, 5, 3, 1, 6, 4, 2, 0);
+  return _mm512_permutexvar_epi64(idx, out);
+}
+
+template <> class Vectorized<BFloat16> {
+private:
+  __m512i values;
+public:
+  using value_type = uint16_t;
+  using size_type = int;
+  static constexpr size_type size() {
+    return 32;
+  }
+  Vectorized() {}
+  Vectorized(__m512i v) : values(v) {}
+  Vectorized(BFloat16 val) {
+    value_type uw = val.x;
+    values = _mm512_set1_epi16(uw);
+  }
+  Vectorized(BFloat16 val1, BFloat16 val2, BFloat16 val3, BFloat16 val4,
+         BFloat16 val5, BFloat16 val6, BFloat16 val7, BFloat16 val8,
+         BFloat16 val9, BFloat16 val10, BFloat16 val11, BFloat16 val12,
+         BFloat16 val13, BFloat16 val14, BFloat16 val15, BFloat16 val16,
+         BFloat16 val17, BFloat16 val18, BFloat16 val19, BFloat16 val20,
+         BFloat16 val21, BFloat16 val22, BFloat16 val23, BFloat16 val24,
+         BFloat16 val25, BFloat16 val26, BFloat16 val27, BFloat16 val28,
+         BFloat16 val29, BFloat16 val30, BFloat16 val31, BFloat16 val32) {
+    values = _mm512_set_epi16(
+        val32.x, val31.x, val30.x, val29.x, val28.x, val27.x, val26.x, val25.x,
+        val24.x, val23.x, val22.x, val21.x, val20.x, val19.x, val18.x, val17.x,
+        val16.x, val15.x, val14.x, val13.x, val12.x, val11.x, val10.x, val9.x,
+        val8.x, val7.x, val6.x, val5.x, val4.x, val3.x, val2.x, val1.x);
+  }
+  operator __m512i() const {
+    return values;
+  }
+  BFloat16& operator[](int idx) = delete;
+  const BFloat16& operator[](int idx) const  = delete;
+  int zero_mask() const {
+    // returns an integer mask where all zero elements are translated to 1-bit and others are translated to 0-bit
+    return _mm512_cmpeq_epi16_mask(values, _mm512_set1_epi16(0));
+  }
+  static Vectorized<BFloat16> loadu(const void* ptr) {
+    return _mm512_loadu_si512(reinterpret_cast<const __m512i*>(ptr));
+  }
+  static Vectorized<BFloat16> loadu(const void* ptr, int16_t count) {
+    __at_align64__ int16_t tmp_values[size()];
+    std::memcpy(tmp_values, ptr, count * sizeof(int16_t));
+    return loadu(tmp_values);
+  }
+  void store(void* ptr, int count = size()) const {
+    if (count == size()) {
+      _mm512_storeu_si512(reinterpret_cast<__m512i*>(ptr), values);
+    } else if (count > 0) {
+      __at_align64__ int16_t tmp_values[size()];
+      _mm512_storeu_si512(reinterpret_cast<__m512i*>(tmp_values), values);
+      std::memcpy(ptr, tmp_values, count * sizeof(int16_t));
+    }
+  }
+  template <int64_t mask>
+  static Vectorized<BFloat16> blend(const Vectorized<BFloat16>& a, const Vectorized<BFloat16>& b) {

```

Comment:
no vector instructions here?

Reply:
The `vec256` implementation uses a macro

https://github.com/pytorch/pytorch/blob/47c566ebb1a0f413b6300cf9700a9cfddb2e5b40/aten/src/ATen/cpu/vec/vec256/intrinsics.h#L14

However, here, I'm accessing the items directly.
I'll check the difference with Godbolt.

Reply:
Accessing items directly is fine, as it doesn't seem to be suboptimal from GodBolt. Moreover, AVX512 doesn't have the corresponding intrinsic.